In [8]:
import pandas as pd

# Load the dataset
file_path = '/Users/minhtao/Downloads/My MacBook/AI and Data Science/Forage/BCGXGenAI/BCGX_Gen_AI/Task 1. Data Extraction and Initial Analysis/Final_report.csv'
df = pd.read_csv(file_path)

# Display the first few rows of the dataframe to understand its structure
df.head()

Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0  2023         211915       72361        411976             205753   
1  2022         198270       72738        364840             198298   
2  2021         168088       61271        333779             191791   
3  2023          96773       14974        106618              43009   
4  2022          81462       12556         82338              36440   

   Cash Flow from Operating Activities    Company  
0                                87582  Microsoft  
1                                89035  Microsoft  
2                                76740  Microsoft  
3                                13256      Tesla  
4                                14724      Tesla

# Columns:
Year: Fiscal year of the data.
Total Revenue: The total revenue for the company in that fiscal year.
Net Income: The net income for the company in that fiscal year.
Total Assets: The total assets for the company in that fiscal year.
Total Liabilities: The total liabilities for the company in that fiscal year.
Cash Flow from Operating Activities: The cash flow from operating activities for the company in that fiscal year.
Company: The name of the company (Microsoft, Apple, or Tesla).

In [9]:
# Calculate year-over-year growth for each financial metric
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income'].pct_change() * 100
df['Assets Growth (%)'] = df.groupby('Company')['Total Assets'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby('Company')['Total Liabilities'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby('Company')['Cash Flow from Operating Activities'].pct_change() * 100

# Display the updated dataframe
df

Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0  2023         211915       72361        411976             205753   
1  2022         198270       72738        364840             198298   
2  2021         168088       61271        333779             191791   
3  2023          96773       14974        106618              43009   
4  2022          81462       12556         82338              36440   
5  2021          53821        5519         62131              30548   
6  2023         383285       96995        352583             290437   
7  2022         394328       99803        352755             302083   
8  2021         365817       94680        351002             287912   

   Cash Flow from Operating Activities    Company  Revenue Growth (%)  \
0                                87582  Microsoft                 NaN   
1                                89035  Microsoft           -6.438902   
2                                76740  Microsoft          -15.222676   
3                                13256      Tesla                 NaN   
4                                14724      Tesla          -15.821562   
5                                11497      Tesla          -33.931158   
6                               110543      Apple                 NaN   
7                               122151      Apple            2.881146   
8                               104038      Apple           -7.230275   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0                    NaN                NaN                     NaN   
1               0.520999         -11.441443               -3.623276   
2             -15.764800          -8.513595               -3.281425   
3                    NaN                NaN                     NaN   
4             -16.147990         -22.772890              -15.273547   
5             -56.044919         -24.541524              -16.169045   
6                    NaN                NaN                     NaN   
7               2.894995           0.048783                4.009820   
8              -5.133112          -0.496945               -4.691095   

   Cash Flow Growth (%)  
0                   NaN  
1              1.659017  
2            -13.809176  
3                   NaN  
4             11.074231  
5            -21.916599  
6                   NaN  
7             10.500891  
8            -14.828368

In [10]:
# Imputing Null/NaN values that result from pct_change calculations with 0
df.fillna(0, inplace=True)
df

Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0  2023         211915       72361        411976             205753   
1  2022         198270       72738        364840             198298   
2  2021         168088       61271        333779             191791   
3  2023          96773       14974        106618              43009   
4  2022          81462       12556         82338              36440   
5  2021          53821        5519         62131              30548   
6  2023         383285       96995        352583             290437   
7  2022         394328       99803        352755             302083   
8  2021         365817       94680        351002             287912   

   Cash Flow from Operating Activities    Company  Revenue Growth (%)  \
0                                87582  Microsoft            0.000000   
1                                89035  Microsoft           -6.438902   
2                                76740  Microsoft          -15.222676   
3                                13256      Tesla            0.000000   
4                                14724      Tesla          -15.821562   
5                                11497      Tesla          -33.931158   
6                               110543      Apple            0.000000   
7                               122151      Apple            2.881146   
8                               104038      Apple           -7.230275   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0               0.000000           0.000000                0.000000   
1               0.520999         -11.441443               -3.623276   
2             -15.764800          -8.513595               -3.281425   
3               0.000000           0.000000                0.000000   
4             -16.147990         -22.772890              -15.273547   
5             -56.044919         -24.541524              -16.169045   
6               0.000000           0.000000                0.000000   
7               2.894995           0.048783                4.009820   
8              -5.133112          -0.496945               -4.691095   

   Cash Flow Growth (%)  
0              0.000000  
1              1.659017  
2            -13.809176  
3              0.000000  
4             11.074231  
5            -21.916599  
6              0.000000  
7             10.500891  
8            -14.828368

In [13]:
df.to_csv('Financial_data_with_growth_rates.csv')

# Summary of Calculated Growth Rates:
Revenue Growth (%): The percentage change in total revenue from one fiscal year to the next.
Net Income Growth (%): The percentage change in net income from one fiscal year to the next.
Assets Growth (%): The percentage change in total assets from one fiscal year to the next.
Liabilities Growth (%): The percentage change in total liabilities from one fiscal year to the next.
Cash Flow Growth (%): The percentage change in cash flow from operating activities from one fiscal year to the next.

In [6]:
# Compute the average growth for each financial metric for each company
avg_growth = df.groupby('Company').mean().drop(columns=['Year', 'Total Revenue', 'Net Income', 'Total Assets', 'Total Liabilities', 'Cash Flow from Operating Activities'])

# Display the average growth for each company
avg_growth


Revenue Growth (%)  Net Income Growth (%)  Assets Growth (%)  \
Company                                                                   
Apple               -1.449710              -0.746039          -0.149388   
Microsoft           -7.220526              -5.081267          -6.651679   
Tesla              -16.584240             -24.064303         -15.771471   

           Liabilities Growth (%)  Cash Flow Growth (%)  
Company                                                  
Apple                   -0.227092             -1.442492  
Microsoft               -2.301567             -4.050053  
Tesla                  -10.480864             -3.614123

In [11]:
avg_growth.to_csv('Average_growth_rates_by_company.csv')

# Apple:
Revenue Growth (%): -1.45%
Net Income Growth (%): -0.75%
Assets Growth (%): -0.15%
Liabilities Growth (%): -0.23%
Cash Flow Growth (%): -1.44%
# Microsoft:
Revenue Growth (%): -7.22%
Net Income Growth (%): -5.08%
Assets Growth (%): -6.65%
Liabilities Growth (%): -2.30%
Cash Flow Growth (%): -4.05%
# Tesla:
Revenue Growth (%): -16.58%
Net Income Growth (%): -24.06%
Assets Growth (%): -15.77%
Liabilities Growth (%): -10.48%
Cash Flow Growth (%): -3.61%

# Summary of Findings:
# Apple:

Relatively stable financial metrics with minimal negative growth, indicating a steady financial position.
# Microsoft:

Significant decline across most financial metrics, particularly in revenue and net income, which might indicate challenges in maintaining its growth trajectory.
# Tesla:

Highest volatility, with substantial negative growth rates across all metrics, reflecting its dynamic and high-risk market environment.

# Implications:
# Apple:

Appears to be in a stable financial state with minor fluctuations, suggesting a conservative approach to growth and risk management.
# Microsoft:

Might need to address underlying issues causing the decline in revenue and net income, potentially through strategic adjustments or innovation.
# Tesla:

Significant declines suggest the need for careful financial management to navigate its high-risk market and maintain sustainability.